File: kits/typescript/src/readiness.ts
// Feature readiness. `blocked` (the environment cannot satisfy it) and `off` (intentionally
// disabled) are EXCLUDED from the overall verdict — a dev box that can never meet the environment
// still reports `ready` (loqu8 invariant #7). ironmcp owns the SHAPE + the verdict semantics;
// the app supplies the feature/lib/data checks (no dart:ffi / node native dep here). Pure.

export type ReadinessStatus = "ready" | "degraded" | "failed" | "blocked" | "off";

export type FeatureReadiness = {
  id: string;
  label: string;
  status: ReadinessStatus;
  requires?: string[];
  details?: string;
  reason?: string;
};

export type LibraryStatus = {
  name: string;
  loaded: boolean;
  symbolsChecked?: number;
  symbolsOk?: number;
  error?: string;
};

export type DataFileStatus = {
  label: string;
  found: boolean;
  path?: string;
};

export type ReadinessReportInput = {
  appVersion: string;
  nativeVersion?: string;
  features?: FeatureReadiness[];
  libs?: LibraryStatus[];
  dataFiles?: DataFileStatus[];
  platform?: Record<string, unknown>;
};

// `id` is the map key under `features`; `label` is a display hint kept out of the
// wire shape so the per-feature value is byte-identical across kits.
function featureToJson(f: FeatureReadiness): Record<string, unknown> {
  const j: Record<string, unknown> = { status: f.status };
  if (f.requires && f.requires.length > 0) j.requires = f.requires;
  if (f.details != null) j.details = f.details;
  if (f.reason != null) j.reason = f.reason;
  return j;
}

// `name` is the map key under `dependencies`. Symbol counts are FFI-specific, so
// they appear ONLY when a probe ran — a non-native dependency (a service, a
// database) carries just `loaded` and an optional `error`.
function libToJson(l: LibraryStatus): Record<string, unknown> {
  const j: Record<string, unknown> = { loaded: l.loaded };
  const checked = l.symbolsChecked ?? 0;
  if (checked > 0) {
    j.symbols_checked = checked;
    j.symbols_ok = l.symbolsOk ?? 0;
  }
  if (l.error != null) j.error = l.error;
  return j;
}

// `label` is the map key under `data_files`.
function dataFileToJson(d: DataFileStatus): Record<string, unknown> {
  const j: Record<string, unknown> = { found: d.found };
  if (d.path != null) j.path = d.path;
  return j;
}

/** A full readiness report. */
export class ReadinessReport {
  readonly appVersion: string;
  readonly nativeVersion?: string;
  readonly features: FeatureReadiness[];
  readonly libs: LibraryStatus[];
  readonly dataFiles: DataFileStatus[];
  readonly platform: Record<string, unknown>;

  constructor(input: ReadinessReportInput) {
    this.appVersion = input.appVersion;
    this.nativeVersion = input.nativeVersion;
    this.features = input.features ?? [];
    this.libs = input.libs ?? [];
    this.dataFiles = input.dataFiles ?? [];
    this.platform = input.platform ?? {};
  }

  /** Overall verdict from the features that COUNT — `blocked`/`off` are excluded (invariant #7).
   *  failed > degraded > ready. */
  get overallStatus(): ReadinessStatus {
    const counted = this.features.filter((f) => f.status !== "blocked" && f.status !== "off");
    if (counted.some((f) => f.status === "failed")) return "failed";
    if (counted.some((f) => f.status === "degraded")) return "degraded";
    return "ready";
  }

  // Ecosystem health-check vocabulary (`status`) + loqu8's map-by-id structure:
  // features/dependencies/data_files are OBJECTS keyed by id/name, so an agent
  // reads `features["<id>"].status` in one hop, there is no list order to keep
  // byte-identical across kits, and duplicate ids cannot hide. `dependencies`
  // (not `libs`) so a server with services rather than native libraries is not
  // misdescribed.
  toJSON(): Record<string, unknown> {
    const j: Record<string, unknown> = { app_version: this.appVersion };
    if (this.nativeVersion != null) j.native_version = this.nativeVersion;
    j.status = this.overallStatus;
    j.features = Object.fromEntries(this.features.map((f) => [f.id, featureToJson(f)]));
    j.dependencies = Object.fromEntries(this.libs.map((l) => [l.name, libToJson(l)]));
    j.data_files = Object.fromEntries(this.dataFiles.map((d) => [d.label, dataFileToJson(d)]));
    j.platform = this.platform;
    return j;
  }
}

File: kits/dart/lib/src/readiness.dart
/// Feature readiness. `blocked` (the environment cannot satisfy it) and `off`
/// (intentionally disabled) are EXCLUDED from the overall verdict — a dev box
/// that can never meet the environment still reports `ready` (loqu8 invariant #7).
enum ReadinessStatus { ready, degraded, failed, blocked, off }

class FeatureReadiness {
  FeatureReadiness({
    required this.id,
    required this.label,
    required this.status,
    this.requires = const [],
    this.details,
    this.reason,
  });

  final String id;
  final String label;
  final ReadinessStatus status;
  final List<String> requires;
  final String? details;
  final String? reason;

  /// The value under `features["<id>"]`. `id` is the map key, so it is not
  /// repeated here; `label` is a display hint kept out of the wire shape so the
  /// per-feature value is byte-identical across every kit.
  Map<String, dynamic> toJson() => {
        'status': status.name,
        if (requires.isNotEmpty) 'requires': requires,
        if (details != null) 'details': details,
        if (reason != null) 'reason': reason,
      };
}

/// A native library check result. ironmcp owns this SHAPE; the actual FFI probe
/// (`DynamicLibrary.open` + `lookup(symbol)`) is supplied by the app, so the kit
/// carries no dart:ffi dependency.
class LibraryStatus {
  LibraryStatus({
    required this.name,
    required this.loaded,
    this.symbolsChecked = 0,
    this.symbolsOk = 0,
    this.error,
  });

  final String name;
  final bool loaded;
  final int symbolsChecked;
  final int symbolsOk;
  final String? error;

  /// The value under `dependencies["<name>"]`. `name` is the map key. The symbol
  /// counts are FFI-specific, so they appear ONLY when a probe ran — a non-native
  /// dependency (a service, a database) just carries `loaded` and an optional
  /// `error`, which keeps `dependencies` meaningful for every kind of server.
  Map<String, dynamic> toJson() => {
        'loaded': loaded,
        if (symbolsChecked > 0) 'symbols_checked': symbolsChecked,
        if (symbolsChecked > 0) 'symbols_ok': symbolsOk,
        if (error != null) 'error': error,
      };
}

class DataFileStatus {
  DataFileStatus({required this.label, required this.found, this.path});

  final String label;
  final bool found;
  final String? path;

  /// The value under `data_files["<label>"]`. `label` is the map key.
  Map<String, dynamic> toJson() =>
      {'found': found, if (path != null) 'path': path};
}

/// A full readiness report. ironmcp owns the shape + the verdict semantics; the
/// app supplies the feature/lib/data checks.
class ReadinessReport {
  ReadinessReport({
    required this.appVersion,
    this.nativeVersion,
    this.features = const [],
    this.libs = const [],
    this.dataFiles = const [],
    this.platform = const {},
  });

  final String appVersion;
  final String? nativeVersion;
  final List<FeatureReadiness> features;
  final List<LibraryStatus> libs;
  final List<DataFileStatus> dataFiles;
  final Map<String, dynamic> platform;

  /// Overall verdict from the features that COUNT — `blocked`/`off` are excluded
  /// (invariant #7). failed > degraded > ready.
  ReadinessStatus get overallStatus {
    final counted = features.where((f) =>
        f.status != ReadinessStatus.blocked && f.status != ReadinessStatus.off);
    if (counted.any((f) => f.status == ReadinessStatus.failed)) {
      return ReadinessStatus.failed;
    }
    if (counted.any((f) => f.status == ReadinessStatus.degraded)) {
      return ReadinessStatus.degraded;
    }
    return ReadinessStatus.ready;
  }

  /// The wire shape. Learned from the ecosystem's health-check convention
  /// (IETF health-check draft / Kubernetes / Spring Actuator all key on `status`)
  /// and from loqu8's map-by-id structure: `features`, `dependencies`, and
  /// `data_files` are OBJECTS keyed by id/name — an agent reads
  /// `features["<id>"].status` in one hop, there is no list order to keep
  /// byte-identical across kits, and duplicate ids cannot hide. `dependencies`
  /// (not `libs`) so a server with services rather than native libraries is not
  /// misdescribed.
  Map<String, dynamic> toJson() => {
        'app_version': appVersion,
        if (nativeVersion != null) 'native_version': nativeVersion,
        'status': overallStatus.name,
        'features': {for (final f in features) f.id: f.toJson()},
        'dependencies': {for (final l in libs) l.name: l.toJson()},
        'data_files': {for (final d in dataFiles) d.label: d.toJson()},
        'platform': platform,
      };
}
